Add session persistence conformance test plan (#4482)
kubernetes-sigs/gateway-api: geps/gep-1619/index.md

| Session Persistence Scoped to Route Rule: HTTPRoute with two rules: Rule 1 (path /a) and Rule 2 (path /b), both with sessionPersistence configured, routing to the same backend Service with multiple pods. | HTTPRoute MUST have Accepted=True in parent status. Each rule MUST independently establish and maintain its own session. Verify with: 1) Send an initial request to /a (no cookie) — the response sets a session cookie binding /a to Pod-A, 2) Send an initial request to /b (no cookie) — the response sets a session cookie independently binding /b to Pod-B (Pod-B may or may not be the same pod as Pod-A), 3) N subsequent requests to /a with the /a session cookie MUST all route to Pod-A, 4) N subsequent requests to /b with the /b session cookie MUST all route to Pod-B. | HTTPRouteSessionPersistence |

> | Session Persistence with cookieConfig.lifetimeType: Permanent and absoluteTimeout: 5min. | HTTPRoute MUST have Accepted=True in parent status. Response Set-Cookie header MUST contain `expiry` attribute. The expiry value MUST correspond to the configured absoluteTimeout duration. Session persistence MUST function correctly until cookie expires. | HTTPRouteSessionPersistence, HTTPRouteSessionPersistenceCookieLifetimeTypePermanent |

Reviewer: ```suggestion
| Session Persistence with cookieConfig.lifetimeType: Permanent and absoluteTimeout: 5min. | HTTPRoute MUST have Accepted=True in parent status. Response Set-Cookie header MUST contain `Expires` or `Max-Age` attribute. The expiry value MUST correspond to the configured absoluteTimeout duration. Session persistence MUST function correctly until cookie expires. | HTTPRouteSessionPersistence, HTTPRouteSessionPersistenceCookieLifetimeTypePermanent |
```

### GRPCRoute Feature Names

* GRPCRouteSessionPersistence - Core feature for session persistence support on GRPCRoute (Extended).
* GRPCRouteSessionPersistenceCookieLifetimeTypePermanent - Permanent Lifetime for cookie-based session persistence on GRPCRoute (Extended)

> > **Note on gRPC limitations**: Unlike HTTP/1.1 clients, gRPC clients do not automatically manage cookies. There is no built-in cookie jar; the client must explicitly extract `Set-Cookie` response headers and include them as `Cookie` metadata in subsequent requests. For this reason, header-based session persistence (`type: Header`) is the more natural and idiomatic mechanism for gRPC workloads. Cookie-based persistence is still technically valid over HTTP/2, but requires explicit client-side handling in tests.

Reviewer: This is a great point - there's a solid argument for having dynamic conformance levels for Cookie/Header depending on HTTPRoute/GRCPRoute which would be an update to the Core GEP.

Would you be willing to remove this from here - and turn it into a github issue that we can discuss and maybe follow up with a GEP update for dynamic conformance levels depending on the type of route?

Author: Opened the issue [here](https://github.com/kubernetes-sigs/gateway-api/issues/4774)

### GRPCRoute Conformance tests

| Description | Outcome | Features |
| :---- | :---- | :---- |
| Simple Cookie-based Session Persistence (GRPCRoute): A GRPCRoute with sessionPersistence configured with type: Cookie (default) on a single backend. The test client MUST explicitly extract the Set-Cookie response header and include it as a Cookie header in subsequent requests. | GRPCRoute MUST have Accepted=True in parent status. First request MUST receive a Set-Cookie response header. Subsequent requests with the cookie header MUST route to the same backend pod. Verify by checking pod identity across N requests (N>=50). | GRPCRouteSessionPersistence |
| Header-based Session Persistence (GRPCRoute): A GRPCRoute with sessionPersistence configured with type: Header on a single backend. | GRPCRoute MUST have Accepted=True in parent status. First request MUST receive a session identity header in the response metadata. Subsequent requests with that header included in request metadata MUST route to the same backend pod. Verify by checking pod identity across N requests (N>=50). | GRPCRouteSessionPersistence |

> | Session Cookie Lifetime (Default) (GRPCRoute): GRPCRoute with sessionPersistence and cookieConfig.lifetimeType: Session (default). | GRPCRoute MUST have Accepted=True in parent status. Cookie MUST NOT contain `expiry` attributes. | GRPCRouteSessionPersistence |

Reviewer: So technically this is incorrect since Header-Based session persistence is Extended (you'll need to add another feature at the top e.g. `GRPCRouteSessionPersistenceHeader`)

There's another conversation to be had here whether GRPCRoute should have inverse conformance levels of HTTPRoute (`Cookie` is Extended while `Header` is Core), but I think we can take that up as a separate issue and focus on writing the conformance tests reflecting the spec as-is.

```suggestion
| Session Cookie Lifetime (Default) (GRPCRoute): GRPCRoute with sessionPersistence and cookieConfig.lifetimeType: Session (default). | GRPCRoute MUST have Accepted=True in parent status. Cookie MUST NOT contain `expiry` attributes. | GRPCRouteSessionPersistenceHeader |
```

Author: can you explain me more on why this is wrong?

Reviewer: Since header is [extended](https://github.com/kubernetes-sigs/gateway-api/blob/8f1414bedc59435930ae5f875cf59f2de87a9aae/apis/v1/shared_types.go#L944), it needs to have its own feature. If it wasn't, there would be no way to say "I support Cookies (core), but I don't support header (extended)" because its wrapped up in `GRPCRouteSessionPersistence`.

Conformance levels are layered. Session persistence itself is Extended (implementations can opt out entirely), but within session persistence, Cookie is Core (you must support it if you claim session persistence) while Header is Extended (you can support session persistence without supporting header-based). So there are two levels of "extended" happening, which is why they need separate features.

So the general rule: if something is extended, it needs its own feature.

Author: thank you for this explanation.

> | Session Cookie Lifetime (Default) (GRPCRoute): GRPCRoute with sessionPersistence and cookieConfig.lifetimeType: Session (default). | GRPCRoute MUST have Accepted=True in parent status. Cookie MUST NOT contain `expiry` attributes. | GRPCRouteSessionPersistence |

Reviewer: ```suggestion
| Session Cookie Lifetime (Default) (GRPCRoute): GRPCRoute with sessionPersistence and cookieConfig.lifetimeType: Session (default). | GRPCRoute MUST have Accepted=True in parent status. Cookie MUST NOT contain `Expires` or `Max-Age` attributes. | GRPCRouteSessionPersistence |
```

- How do we provide a standard way to communicate that an implementation does not support Route Rule API?
    - Do we want something conceptually similar to the `IncompatibleFilters` reason?

## Conformance Details

###  Feature Names

* HTTPRouteSessionPersistence - Base feature for session persistence support
* HTTPRouteSessionPersistenceCookie - Cookie-based session persistence (Core)
* HTTPRouteSessionPersistenceCookieLifetimeTypePermanent - Permanent Lifetime for cookie-based session persistence (extended)

> 

Reviewer: We should probably drop a note that this is still WIP, so folks aren't confused why it's not complete.

```suggestion

**Note**: Conformance tests are still WIP, and additional test coverage will be added in future PRs.
```

- How do we provide a standard way to communicate that an implementation does not support Route Rule API?
    - Do we want something conceptually similar to the `IncompatibleFilters` reason?

> 

Reviewer 2: This is not directly relevant to the changes in this PR, but before graduating the feature to Standard it would be good to have clarity about field names. I see `sessionPersistence.sessionName` and `sessionPersistence.name` used interchangeably in the GEP.

In `sessionPersistence.sessionName` we are repeating "session", which is generally discouraged. It also makes a misleading impression that this feature is about establishing some "session".

To be in line with the comment in the code snippet, it should ideally be `sessionPersistence.tokenName`, but that's probably overly technical, so perhaps `sessionPersistence.name` is close enough.

## Conformance Details

> ###  Feature Names

Reviewer: I'm good with delaying `BackendTrafficPolicy`, as that is up in the air. For `GRPCRoute`, we can add it here together, but I think it's better to add it in a follow-up PR.

* HTTPRouteSessionPersistence - Base feature for session persistence support
* HTTPRouteSessionPersistenceCookie - Cookie-based session persistence (Core)
* HTTPRouteSessionPersistenceCookiePermanentLifetime - Permanent Lifetime for cookie-based session persistence (extended)

### Conformance tests

| Description | Outcome | Features |
| :---- | :---- | :---- | 
| Simple Cookie Session Persistence: An HTTPRoute with sessionPersistence configured with type: Cookie (default) on a single backend in gateway-conformance-infra namespace. | HTTPRoute MUST have Accepted=True in parent status. First request MUST receive a Set-Cookie header in response. Subsequent requests with the cookie MUST route to the same backend pod. Verify by checking pod identity across N requests (N>=50). | HTTPRouteSessionPersistence, HTTPRouteSessionPersistenceCookie |

> | Session Cookie Lifetime (Default): HTTPRoute with sessionPersistence and cookieConfig.lifetimeType: Session (default). | HTTPRoute MUST have Accepted=True in parent status. Cookie MUST NOT contain `expiry` attributes. | HTTPRouteSessionPersistence, HTTPRouteSessionPersistenceCookie |

Reviewer 2: What about the case of two paths in one rule?
```
kind: HTTPRoute
metadata:
  name: routeX
spec:
  rules:
  - matches:
    - path:
      value: /a
    - path:
      value: /b
    backendRefs:
    - name: servicev1
    sessionPersistence:
      name: session
```

should they be shared? I guess it may be technically challenging because there can be only one path specified in Set-Cookie ([link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#pathpath-value)). Do we want to specify the behavior here or leave it as implementation specific?

Reviewer: Agree, I want to update this in the GEP, it doesn't specify how or if the two paths on the same rule should share. I think it's likely we should prescribe sharing in the GEP, and then we can follow up with a new conformance test. For now, the current tests are fine.
